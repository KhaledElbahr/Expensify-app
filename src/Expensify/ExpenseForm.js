import React, { Component } from 'react'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'
import { v4 as uuid } from 'uuid'

export default class ExpenseForm extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createAt) : moment(),
      calendarfocused: false,
      title: props.expense ? 'Update' : 'Add',
      error: ''
    }
  }

  getValue = e => e.target.value

  onDescriptionChange = e => {
    const description = this.getValue(e)
    this.setState(() => ({ description }))
  }

  onNoteChange = e => {
    const note = this.getValue(e)
    this.setState(() => ({ note }))
  }

  onAmountChange = e => {
    const amount = this.getValue(e)
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }

  onDateChange = createdAt => this.setState(() => ({ createdAt }))

  onFocusChange = calendarfocused => {
    this.setState(() => ({ calendarfocused }))
  }

  onSubmit = e => {
    e.preventDefault()

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount' }))
    } else {
      this.setState(() => ({ error: '' }))
      this.props.onSubmit({
        description: this.state.description,
        note: this.state.note,
        amount: parseFloat(this.state.amount, 10) * 100,
        createAt: this.state.createdAt.valueOf()
      })
    }
  }

  render () {
    return (
      <form className='form' onSubmit={this.onSubmit}>
        {this.state.error && <p className='form__error'>{this.state.error}</p>}
        <input
          type='text'
          className='text-input'
          placeholder='Description'
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type='text'
          className='text-input'
          value={this.state.amount}
          onChange={this.onAmountChange}
          placeholder='Amount'
        />
        <div className='input-control'>
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarfocused.focused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            id={uuid()} // PropTypes.string.isRequired,
            showClearDate={true}
            isOutsideRange={() => false}
          />
        </div>
        <textarea
          value={this.state.note}
          className='textarea'
          onChange={this.onNoteChange}
          placeholder='Add a note for your expense (optional)'
        ></textarea>
        <div className="bototn--group">
          <button className='button'>{this.state.title} Expense</button>
          {
            this.props.expense 
            && 
            <button className="button button--danger" onClick={this.props.onRemove}>Delete Expense</button>
          }
        </div>
      </form>
    )
  }
}
