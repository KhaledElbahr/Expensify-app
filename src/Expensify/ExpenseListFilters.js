import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  setTextFilter,
  sortByDateFilter,
  sortByAmountFilter,
  setStartDateFilter,
  setEndDateFilter
} from './../redux/actions/filters';
import { DateRangePicker } from 'react-dates';
import { v4 as uuid } from 'uuid';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDateFilter(startDate);
        this.props.setEndDateFilter(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }

    onTextChange = (e) => {
      this.props.setTextFilter(e.target.value)
    }

    onSortChange = (e) => {
      if (e.target.value === 'date') {
        this.props.sortByDateFilter()
      } else if (e.target.value === 'amount') {
        this.props.sortByAmountFilter()
      }
    }

    render () {
    return (
      <div>
        <input
          type='text'
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />

        <select
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}
        >
          <option value='date'>Date</option>
          <option value='amount'>Amount</option>
        </select>

        <DateRangePicker
        startDate={this.props.filters.startDate}
        endDate={this.props.filters.endDate}
        // startDateId={uuid()}
        // endDateId={uuid()}
        onDatesChange={this.onDatesChange}
        focusedInput={this.state.calendarFocused}
        onFocusChange={this.onFocusChange}
        showClearDates={true}
        numberOfMonths={1}
        isOutsideRange={() => false}
        />
      </div>
    )
  }
}

// const ExpenseListFilters = props => (
//   <div>
//     {/* {props.filters ? (
//             <>
//                 <p>{props.filters.sortBy}</p>
//                 <p>{props.filters.text}</p>
//                 {props.filters.startDate && <p>{props.filters.startDate}</p>}
//                 {props.filters.endDate && <p>{props.filters.endDate}</p>}
//             </>
//             ) : null
//         } */}
//     <input
//       type='text'
//       value={props.filters.text}
//       onChange={e => {
//         props.dispatch(setTextFilter(e.target.value))
//       }}
//     />

//     <select
//       value={props.filters.sortBy}
//       onChange={e => {
//         if (e.target.value === 'date') {
//           props.dispatch(sortByDateFilter())
//         } else if (e.target.value === 'amount') {
//           props.dispatch(sortByAmountFilter())
//         }
//       }}
//     >
//       <option value='date'>Date</option>
//       <option value='amount'>Amount</option>
//     </select>
//   </div>
// )

const mapStateToProps = state => ({
  filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDateFilter: () => dispatch(sortByDateFilter()),
  sortByAmountFilter: () => dispatch(sortByAmountFilter()),
  setStartDateFilter: (startDate) => dispatch(setStartDateFilter(startDate)),
  setEndDateFilter: (endDate) => dispatch(setEndDateFilter(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)
