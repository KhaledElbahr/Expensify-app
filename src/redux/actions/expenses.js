import { v4 as uuid } from 'uuid';
import db from './../../firebase/firebase';
import { push, ref } from "firebase/database";
// actions
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

// Add Expense Data
export const addExpenseData = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createAt = 0
    } = expenseData;
    const expense = { description, note, amount, createAt };

    return push(ref(db, 'expenses/'), expense)
    .then(data => {
      dispatch(addExpense({
        id: data.key,
        ...expense       
      }));
    })
    .catch((e) => console.log('This failed', e))
  }
}

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})
