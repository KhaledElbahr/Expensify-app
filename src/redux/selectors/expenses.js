import moment from 'moment';

// Get Visibile Expenses
export default (expenses, {
    text,
    sortBy,
    startDate,
    endDate
}) => {
    return expenses.filter((expense) => {
        // const endDateMatch = typeof endDate !== 'number' || expense.createAt <= endDate;
        // const startDateMatch = typeof startDate !== 'number' || expense.createAt >= startDate;
        const createdAtMoment = moment(expense.createAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if(sortBy === 'date') {
            return a.createAt < b.createAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
    })
}
