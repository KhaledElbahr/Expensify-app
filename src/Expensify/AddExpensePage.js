import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addExpenseData } from './../redux/actions/expenses';
import ExpenseForm from './ExpenseForm';
import { useNavigate } from "react-router-dom";

export class AddExpensePage extends React.Component {
    constructor(props) {
        super(props);
    }

    onSubmit = (expense) => {
        this.props.addExpenseData(expense);
        this.props.navigate('/dashboard');
    }

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="container">
                        <h2 className="page-header__title">Add Expense</h2>
                    </div>
                </div>
                <div className="container">
                    <ExpenseForm onSubmit={this.onSubmit} />
                </div>
            </div>
        )
    }
}

// const AddExpensePage = (props) => {
//     const navigate = useNavigate();

//     return (
//     <div>
//         <h2>Add Expense</h2>
//         <ExpenseForm
//         onSubmit={(expense) => {
//             // props.dispatch(addExpense(expense));
//             props.onSubmit(expense);
//             navigate("/");
//         }}
//         />
//     </div>
// )}

const mapDispatchToProps = (dispatch) => {
    const navigate = useNavigate();
    
    return {
        addExpenseData: (expense) => dispatch(addExpenseData(expense)),
        navigate: (toPath) => navigate(toPath)
    }
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);