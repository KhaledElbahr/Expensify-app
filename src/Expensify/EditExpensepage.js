import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpenseData } from '../redux/actions/expenses';
import { useNavigate } from 'react-router-dom';

export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props);
    }

    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.navigate('/dashboard');
    }

    onRemoveExpense = () => {
        this.props.removeExpenseData({ id: this.props.expense.id });
        this.props.navigate('/dashboard');
    }

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="container">
                        <h2 className="page-header__title">Edit Expense</h2>
                    </div>
                </div>
                <div className="container">
                    <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick={this.onRemoveExpense}>Remove Expense</button>
                </div>
            </div> 
        )
    }
}

// const EditExpensePage = (props) => (
//     <div>
//         <ExpenseForm 
//         expense={props.expense}
//         onSubmit={(expense) => {
//             props.dispatch(editExpense(props.expense.id, expense));
//             props.history.push('/');
//         }} 
//         />
//         <button onclick={() => (
//             props.dispatch(removeExpense({ id: props.expense.id }));
//             props.history.push('/');
//         )}>Remove</button>
//     </div>
// );

const mapStateToProps = (state) => {
    return {
        expense: state.expenses.find((expense) => expense.id === window.location.href.split('/')[4])
    }
}

const mapDispatchToProps = (dispatch) => {
    const navigate = useNavigate();
    
    return {
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        removeExpenseData: (id) => dispatch(removeExpenseData(id)),
        navigate: (toPath) => navigate(toPath),
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
