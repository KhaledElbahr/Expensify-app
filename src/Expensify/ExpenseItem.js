import React from 'react';
import { removeExpense } from './../redux/actions/expenses';
import { Link } from 'react-router-dom';
import moment from "moment";
import numeral from "numeral";

export const ExpenseItem = ({
    id, 
    note, 
    description, 
    amount, 
    createAt,
    dispatch
}) => (
    <Link className="list-item" to={`/edit-expense/${id}`}>
        <div>
            <h3 className="list-item__title">{description}</h3>
            <small className="list-item__sub-title">{moment(createAt).format('D MMM YYYY')}</small>
        </div>
        <div>
            <h3 className="list-item__data">{numeral(amount / 100).format('$0,0.00')}</h3> 
        </div>
    </Link>
)

export default (ExpenseItem);