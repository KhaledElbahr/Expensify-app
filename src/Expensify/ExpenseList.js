import React from 'react';
import { connect } from 'react-redux';
import ExpenseItem from './ExpenseItem';
import selectedExpenses from '../redux/selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        <h2>Expense List</h2>
        {props.expenses && props.expenses.length !== 0 ? 
        (<>
            {props.expenses.map((expense) => (
            <ExpenseItem key={expense.id} {...expense} /> 
            ))}
        </>) : (<p>There is no Expenses!!</p>)
        }
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectedExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList);