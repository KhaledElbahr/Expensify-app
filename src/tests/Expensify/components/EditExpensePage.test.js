import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { EditExpensePage } from './../../../Expensify/EditExpensepage';
import expenses from './../../fixtures/expenes';

let editExpense, removeExpense, navigate, wrapper;

beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    navigate = jest.fn();
    wrapper = shallow(<EditExpensePage
        editExpense={editExpense} 
        removeExpense={removeExpense}
        navigate={navigate}
        expense={expenses[1]}
        />);
});

test('should render EditExpensePage correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
});

test('should handle editExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(navigate).toHaveBeenLastCalledWith('/dashboard');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});

test('should handle removeExpense', () => {
    wrapper.find('button').simulate('click');
    expect(navigate).toHaveBeenLastCalledWith('/dashboard');
    expect(removeExpense).toHaveBeenLastCalledWith({
        id: expenses[1].id
    });
});