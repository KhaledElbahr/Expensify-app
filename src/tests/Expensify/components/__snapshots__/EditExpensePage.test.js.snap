// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="container"
    >
      <h2
        className="page-header__title"
      >
        Edit Expense
      </h2>
    </div>
  </div>
  <div
    className="container"
  >
    <ExpenseForm
      expense={
        {
          "amount": 350,
          "createAt": -345600000,
          "description": "Water",
          "id": "2",
          "note": "This is some water",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
